include/seqan3/io/alignment_file/format_sam.hpp

```diff
@@ -344,9 +638,406 @@ class alignment_file_format_sam
+     *
+     * \throws seqan3::parse_error if any unexpected character or format is encountered.
+     *
+     * \details
+     *
+     * Reading the header format is done according to the official
+     * [SAM format specifications](https://samtools.github.io/hts-specs/SAMv1.pdf).
+     *
+     * The function throws a seqan3::parse_error if any unknown tag was encountered. It will also fail if the format is
+     * not in a correct state (e.g. required fields are not giving), but throwing might occur downstream of the actual
+     * error.
+     */
+    template <typename stream_view_type>
+    void read_header(stream_view_type && stream_view, std::unique_ptr<alignment_file_header> & header_ptr)
+    {
+        alignment_file_header & hdr{*header_ptr};
+
+        auto parse_tag_value = [&stream_view, this] (auto & value) // helper function to parse the next tag value
+            {
+                detail::consume(stream_view | view::take_until_or_throw(is_char<':'>)); // skip tag name
+                ++begin(stream_view);                                                   // skip ':'
+                read_field(stream_view | view::take_until_or_throw(is_char<'\t'> || is_char<'\n'>), value);
+            };
+
+        // @HQ line
+        // -------------------------------------------------------------------------------------------------------------
+        parse_tag_value(hdr.format_version); // parse required VN (version) tag
+
+        // The SO, SS and GO tag are optional and can appear in any order
+        while (is_char<'\t'>(*begin(stream_view)))
+        {
+            ++begin(stream_view); // skip tab
+            std::string * who = &hdr.grouping;
+
+            if (is_char<'S'>(*begin(stream_view)))
+            {
+                ++begin(stream_view); // skip S
+
+                if (is_char<'O'>(*begin(stream_view)))          // SO (sorting) tag
+                    who = &hdr.sorting;
+                else if (is_char<'S'>(*begin(stream_view)))     // SS (sub-order) tag
+                    who = &hdr.subsorting;
+                else
+                    throw parse_error(std::string("Illegal SAM header tag: S" + (*begin(stream_view))));
+            }
+            else if (!is_char<'G'>(*begin(stream_view)))        // GO (grouping) tag
+            {
+                throw parse_error(std::string("Illegal SAM header tag in @HG starting with:" + (*begin(stream_view))));
+            }
+
+            parse_tag_value(*who);
+        }
+        ++begin(stream_view);  // skip newline
+
+        // The rest of the header lines
+        // -------------------------------------------------------------------------------------------------------------
+        while (is_char<'@'>(*begin(stream_view)))
+        {
+            ++begin(stream_view); // skip @
+
+            if (is_char<'S'>(*begin(stream_view)))              // SQ (sequence dictionary) tag
```

Comment:
So I could do an `@S` in the header and it would not throw? Because it should be `@SQ` and not `@S`

Reply:
Yes but there is no convenient way to assert this without refactoring the code and I think we are not responsible to check the header for strict correctness.. We could change this in the future.